ArC and Vert.x: introduce `InvokerFactoryBuildItem` and use it in Vert.x event bus consumers

Ladicek opened the pull request:
The last commit fixes #26728

Ladicek committed:
ArC: fix method invoker argument transformers in case of primitive types

Ladicek committed:
ArC: add InvokerFactoryBuildItem

Ladicek committed:
ArC: fix method invokers async CC.release() in presence of a return value transformation

Ladicek committed:
Vert.x: use method invokers for event bus consumers

Ladicek committed:
Vert.x: support Kotlin suspend functions in event bus consumers

Ladicek commented on extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/InvokerFactoryBuildItem.java:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Holds an ArC {@link InvokerFactory}. May only be used in the bean registration phase,
+ * observer registration phase, and validation phase; afterwards, any registered invokers
+ * are ignored. TODO this should be enforced somehow?
```
Any ideas how to enforce this [using build items probably]? I come up empty :shrug:

mkouba replied:
I don't think you can enforce the usage of a build item, except for throwing an exception... Well, you can store the `BeanDeployment`  reference instead of `InvokerFactory` and throw an exception directly from the `BeanDeployment#getInvokerFactory()`.

I.e. something like:
```java
public InvokerBuilder createInvoker(BeanInfo bean, MethodInfo method) {
        return beanDeployment.getInvokerFactory().createInvoker(bean, method);
}
```

Ladicek replied:
Hmm yeah, that makes sense. Thanks!

Ladicek replied:
Done.

Ladicek commented:
Probably easiest to review each commit separately.

cescoffier commented on extensions/vertx/deployment/src/test/kotlin/io/quarkus/vertx/kotlin/KotlinSuspendConsumeEventTest.kt:
```diff
@@ -0,0 +1,133 @@
+package io.quarkus.vertx.kotlin
```
Could it introduce an issue to depend on Kotlin here?

cescoffier commented on extensions/vertx/runtime/pom.xml:
```diff
@@ -46,6 +46,10 @@
         </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-vertx-kotlin</artifactId>
```
That's not great. Now everything depends on Kotlin.

Ladicek replied:
It's just a tiny artifact with 3 classes and transitive dependencies on artifacts that are already present, but I agree it is not nice. I took this over from the Scheduler extension without thinking too much about it. I'll ponder over it over the weekend.

mkouba replied:
I think that it's fine. It does not really bring the Kotlin sdk in (it's an optional dependency)...

Ladicek replied:
Added one more commit (to be squashed if you guys like it :-) ) that moves the dependency to the `quarkus-kotlin` module. Probably not very nice either, but should be better than a dependency on Kotlin in the `quarkus-vertx` module, I guess.

cescoffier reviewed (changes requested):
We need to rework the dependency on Kotlin. Main extensions should not depend on it.

mschorsch commented:
Thank you @Ladicek  for taking the time to improve the Kotlin support in Quarkus 👍

mkouba commented on extensions/vertx/deployment/src/main/java/io/quarkus/vertx/deployment/EventConsumerBusinessMethodItem.java:
```diff
@@ -1,33 +1,52 @@
-    private final MethodInfo method;
+    private final boolean blockingAnnotation;
+    private final boolean runOnVirtualThreadAnnotation;
+    private final boolean splitHeadersBodyParams;
```
Would you care to add some javadoc that would explain what this actually means?

Ladicek replied:
Will do.

Ladicek replied:
Done.

Ladicek commented:
Rebased and added one more commit (again, to be squashed) that turns the Kotlin support into an extension which is automatically present when both `quarkus-kotlin` and `quarkus-vertx` are present (due to https://quarkus.io/guides/conditional-extension-dependencies). Seems like the best way forward to me.

Ladicek commented:
Rebased and squashed the last 3 commits.

manovotn reviewed (approved):
Went over ArC parts and those LGTM

mkouba commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanProcessor.java:
```diff
@@ -192,6 +192,8 @@ public List<Resource> generateResources(ReflectionRegistration reflectionRegistr
             ExecutorService executor)
             throws IOException, InterruptedException, ExecutionException {
 
+        beanDeployment.markAsResourcesGenerated();
```
If it's called here the method should be named `resourceGenerationStarted()` or something like that. But that's only nitpicking ;-).

gsmet replied:
@Ladicek I let you decide if you want to act on it. I'm fine with you merging and adjusting later (or not) if you prefer and want to avoid further conflicts.

Ladicek replied:
I thought about this (= `InvokerFactoryBuildItem` lifecycle) again and couldn't find a better solution. So I just applied @mkouba's suggestion.

mkouba reviewed (approved):
Looks good!

geoand commented:
Very cool!

So I assume we should start using invokers wherever possible?

Ladicek commented:
That sounds reasonable, but at the same time, there are occasions where custom bytecode generation is more straightforward. I also vaguely recall that I couldn't use invokers in RESTEasy Reactive, for some reason.
